### PR TITLE
Change SVM._fit_proba implementation

### PR DESCRIPTION
**Pull request by Alexsandruss**
### Description 
_Add a comprehensive description of proposed changes_

Fixes # - _issue number(s) if exists_

- [ ] I have reviewed my changes thoroughly before submitting this pull request.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation to reflect the changes, if necessary (updated in # - _add PR number_)
- [ ] The unit tests pass successfully.
- [ ] I have run it locally and tested the changes extensively.
- [ ] I have resolved any merge conflicts that might occur with the base branch.
- [ ] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_
- [ ] I have added a respective label(s) to PR if I have a permission for that.

### CODE REVIEW

**Comment by samir-nasibli** [comment]
Thank you @Alexsandruss !
For better review please provide some inputs in the description:
* Justification why it is needed?
* What do we achieve with these changes?
* What will we change for the end user?

**Review by ethanglaser — APPROVED** [review]

**Review by samir-nasibli — CHANGES_REQUESTED** [review]
Please address comment above

**Comment by icfaust** [comment]
I think there is a causation issue with switching to 'prefit' on CalibratedClassifierCV (https://stats.stackexchange.com/a/263411), which can explain possible overfitting. Sklearn's SVM is based on the libsvm package (https://www.csie.ntu.edu.tw/~cjlin/libsvm/faq.html#/Q06:_Probability_outputs) which also uses a cross-validation strategy.  Looking at the libsvm source code, the probability metrics are calculated via a default 5 fold stratified Kfold strategy, where the training data fit normally. and then the data is split for the probability calculations into the stratified 5-folds (https://github.com/cjlin1/libsvm/blob/master/svm.cpp#L2067 (regression) https://github.com/cjlin1/libsvm/blob/master/svm.cpp#L1903 (classification)). Our current implementation faithfully reproduces this, but the use of the scikit-learn calibrated cross-validators always seemed like a patch-over to the real solution: implementing a cross-validation strategy in oneDAL.

**Review by icfaust — COMMENTED** [review]
I won't put a changes requested on the PR, but I'd like to discuss my comment above before any action on the PR is taken (i.e. merge or close).

**Comment by ethanglaser** [comment]
> I won't put a changes requested on the PR, but I'd like to discuss my comment above before any action on the PR is taken (i.e. merge or close).

Yeah I maybe prematurely approved this as I didn't realize offline discussion would be necessary, let's make sure we are all aligned on this before merge

**Comment by Alexsandruss** [comment]
@icfaust @ethanglaser

There are three options about acceleration of SVM classification with probabilities:

- Enable `n_jobs` in `CalibratedClassifierCV` to parallelize training of SVMs, but this way is not stable across all cases due to parallelization backend specifics: `loky` backend has significant startup penalty resulting in long first run while `multiprocessing` and `threading` don't give consistent speedup.
- Leave current approach because it's mathematically correct.
- **Switch to prefit estimator in `CalibratedClassifierCV` for now and optimize calibration on oneDAL side later (gives 2-3x speedup consistently across all cases)**. While this perf. fix is not mathematically correct as calibration data should be separate from fitting data, real impact to quality metrics of models trained in benchmarks is low (except logloss which is already known to be different from stock sklearn).

I suggest to go with last one.

Aggregated benchmark results for this PR:
image

**Comment by ethanglaser** [comment]
Thank you for providing additional context and for encouraging discussion on this. I don’t have objections to the last suggested option. Let's make sure a ticket is created for OneDAL implementation, rebase and re-run CI. But I stand by my initial approval at this point.

**Comment by samir-nasibli** [comment]
@Alexsandruss does this fix original issue on #621
Is it observed yet?
I didn't get what was the root cause of the original issue. Could you please bring clarification in description?

**Comment by Alexsandruss** [comment]
@samir-nasibli, description already covers it: prefit estimator is faster than cross-validation.

**Comment by samir-nasibli** [comment]
> @samir-nasibli, description already covers it: prefit estimator is faster than cross-validation.

@Alexsandruss I found a partial answer not in the description of the PR, but in this comment https://github.com/intel/scikit-learn-intelex/pull/1879#issuecomment-2196646328

As I understand it, this is not a fix, but rather a workaround, taking into account mathematical correctness.

But I have to ask, why in this case are we worse than standard sklearn? What is the real root cause? bottle neck? I still don’t understand what it is so that fixes can be addressed specifically against it.

Usually in such cases, if there is no perf, we fallback to the original sklearn.

I also see a valid comment from @icfaust, and assumed that the fix would be approved and merge after the quorum agreed. I didn't understand the rush.

But anyway, thanks for the updates.

**Comment by icfaust** [comment]
@Alexsandruss please send me the ticket for the follow-up development based off of this PR.
